### PR TITLE
feat(release): publish the desktop GUI as an unsigned alpha artifact (#740)

### LINKED ISSUE #740: PRD: publish the desktop GUI for macOS, Windows and Linux on every release (unsigned alpha)
**Placeholder PRD** — intent and measured constraints. Detail gets written when work starts.

> **Split note (2026-08-30):** this issue was originally "publish an alpha artifact" and covered signing too. It is now **publishing only — unsigned**. Making the artifacts trusted (Apple notarization, Windows Authenticode/SmartScreen) is #757, because the two have different blockers: this one is pure engineering with nothing to procure, that one is mostly waiting on certificates.

## Intent

Every release publishes the desktop GUI (PRD #176, `desktop/`) as an **alpha** artifact for **macOS, Windows and Linux**, so it can be installed the way a user would rather than built from source.

Unsigned is acceptable and expected here. Users clear the OS warning manually (`xattr -dr com.apple.quarantine` on macOS, "More info → Run anyway" on Windows); removing that friction is #757.

## Where it stands today

`release.yml` contains no reference to `desktop`, `tauri`, `pnpm`, `bundle`, `dmg` or `.app`. Its `build` job is a four-target matrix running `cargo build --release --target …`, which uses cargo's default target selection — the root package alone — so `desktop/src-tauri` is not even compiled there.

The matrix **already runs on `macos-latest`** for the two darwin targets and `ubuntu-latest` for the Linux ones, so runners are in place.

The bundling machinery exists but is macOS-shaped:

- `pnpm bundle:app` → `tauri build --config src-tauri/tauri.bundle.conf.json --bundles app`. `app` is the **macOS `.app`** target only — not `msi`/`nsis`, not `deb`/`rpm`/`appimage`.
- `desktop/scripts/prepare-sidecar.sh` stages the matching `dot-agent-deck` as a Tauri sidecar so the bundle carries its own daemon. It copies `target/<triple>/<dir>/dot-agent-deck` to `dot-agent-deck-<triple>` with **no `.exe` anywhere**, so it fails on Windows on a missing source file. It is also `#!/bin/sh`.

Compiling is fine on all three — CI proves it on #416 (`build` on Ubuntu, `build-macos`, `build-windows` all green). Only bundling is missing.

## Per-platform effort (the three are very unequal)

| Platform | Bundle | Notes |
| --- | --- | --- |
| **Linux** | `.deb`, `.AppImage` | Easiest. No trust gate at all. Runner already installs Tauri's system deps for the `build` job. |
| **macOS** | `.app` (zipped) or `.dmg` | `--bundles app` produces a directory, not a file — needs zipping for a release asset. `macos-latest` is arm64; decide explicitly whether x86_64 is also built. |
| **Windows** | `.msi` / NSIS | Needs the `prepare-sidecar.sh` `.exe` fix first. **Also ships degraded until #754**: live workflow launch is disabled on Windows, and that guard is itself buggy — once #754 lands the block should follow the *daemon's* platform, so a Windows GUI against a Linux daemon would be fully functional. |

## Also to settle

- **Version.** `tauri.conf.json` hardcodes `version: "0.1.0"`, unrelated to the project's `0.38.x`. Inject the real version rather than shipping a number that means nothing.
- **Opt-in.** Gate behind a `workflow_dispatch` input or tag suffix so it cannot slow or break an ordinary tag.
- **Prerelease.** Mark the assets prerelease so they are not mistaken for supported builds.
- **Desktop platform is independent of daemon platform.** The GUI is a client and can ship where the daemon is not yet supported. The daemon's own Windows track is PRD #42 and #163 (**done**) and #164 (**not started**) and must not gate GUI packaging.

## Suggested first slice

**macOS arm64, unsigned, prerelease, undocumented** — roughly half a day, and enough to start dogfooding immediately. Add Linux next (no trust gate). Windows last, after #754.

## Related

- #757 — making these artifacts trusted (signing, notarization, SmartScreen)
- #754 — the platform guard that keeps Windows degraded
- #746 — the logo; matters for distribution, not for a private alpha
- PRD #176 — M5.1 is the milestone this picks up

### PR DESCRIPTION
**Pull request by vfarcic**
Implements PRD #740. The PRD document and the implementation are in this one PR by request, rather than split across a docs PR and a code PR.

## What this does

Every tag now publishes the desktop GUI alongside the CLI binaries — a `.dmg` for macOS arm64 and a `.deb` for Linux x86_64 — as clearly-labelled unsigned alpha assets with their own `checksums-desktop-alpha.txt`, plus a fixed unsigned-alpha section appended to the release notes.

## Cross-version contract (CLAUDE.md rule 12)

**Did this change the TUI↔daemon contract? No.** No daemon, protocol, orchestration or hook code is touched — this branch changes **zero files under `src/`**. `PROTOCOL_VERSION` is unmoved and no field changes meaning, so there is no wire break and no semantic break. Packaging and CI only, which is **feature → patch** under the `0.x` policy in `docs/develop/versioning.md`. Fragment: `changelog.d/740.feature.md`.

## The design worth reviewing

The topology *is* the guarantee. `desktop-bundle` hangs off `prepare` rather than `build`, and **`finalize` deliberately does not list it in `needs:`** — so the GitHub Release and its five CLI assets are created on exactly today's ~6-minute path whether the bundler succeeds, fails, or is skipped. A bundler failure turns the run red while leaving the release green and complete; `continue-on-error` was rejected because it would make a silently missing artifact indistinguishable from a healthy release.

An opt-in `workflow_dispatch` trigger (which the issue suggested) was rejected too: something a human must remember to fire gets fired for two releases and then never, which is how an alpha artifact ends up six versions stale and worse than none. `skip_desktop` exists as an escape hatch, not an opt-in.

Three single lines carry that guarantee and all three read as housekeeping, so `xtask/linkage-check/src/release_workflow_wiring.rs` asserts them and **each assertion was mutation-tested**. Nothing else can check a workflow that only fires on a tag — by the time a bad edit is observable, a release has already gone out wrong.

One interaction the PRD did not anticipate: `finalize`'s `download-artifact` needed a `pattern:`. `merge-multiple: true` flattens every artifact of the run into one directory, so desktop bundles would have raced into `dist/` depending on job ordering, been swept into the release by the `dist/dot-agent-deck-*` glob, and fed to `task checksums`' `shasum -a 256 dot-agent-deck-*`, which cannot cope with the space in `Agent Deck_0.38.1_amd64.deb`.

## Windows is deferred, and the dependency is #741 — not #164

The issue framed this as blocked on the `.exe` fix plus #754. The decisive question is simpler: **what daemon would a Windows GUI talk to?** No Windows daemon is published, and a Windows GUI cannot reach a remote Linux one either — the IPC transport is `cfg`-dispatched, a named pipe on Windows and a Unix socket everywhere else, and the desktop crate contains no remote code at all. So a Windows alpha today is an app that can connect to nothing.

That makes the dependency **#741** (connect the GUI to any daemon, anywhere), not #164 — #164 would unlock only the local Windows story, which is not the shape wanted. #741's placeholder body does not currently address Windows transport; that decision belongs there.

The `.exe` fix lands anyway as groundwork, covered by `xtask/linkage-check/src/sidecar_staging.rs`. Both halves were confirmed load-bearing by reverting each alone, and they fail with different messages rather than one indistinguishable red.

## Verified by a real bundle, not only by reasoning

A real Linux `tauri build` was run. It produced `Agent Deck_0.38.1_amd64.deb`, which `dpkg-deb -f` reports as `Package: agent-deck`, `Version: 0.38.1` — the injected version reached the package metadata, not just the filename, and the `0.1.0` literal never appears. `dpkg-deb -c` confirms the bundle carries **both** `usr/bin/dot-agent-deck-desktop` and the sidecar daemon `usr/bin/dot-agent-deck`. The collect-and-rename step was then run against that real output and handled the space correctly.

That run also caught a bug that would otherwise have shipped: the bundle generates `desktop/src-tauri/tauri.release.conf.json` in the working tree, which unignored would dirty `git status` on every local bundle until someone committed a version number into the one file whose whole point is never carrying one. Now gitignored.

## AppImage was planned and is dropped, on evidence

It failed (`failed to run linuxdeploy`) while the `.deb` from the same invocation succeeded. Flakiness alone would only justify a retry; what justifies removal is what the failure exposed. Producing an AppImage downloads **five** third-party artifacts at bundle time, two of them shell scripts from `raw.githubusercontent.com/…/master/…` — a **mutable ref**. A release pipeline should not execute whatever happens to be on someone else's default branch at tag time while producing software users install. Re-adding it is one word in the matrix once those inputs are pinned.

## What is NOT verified

**Everything macOS.** No Apple hardware was available, so the `.dmg` path and whether `tauri-bundler` synthesizes a missing `.icns` from the 512×512 PNG (there is no `.icns` in the tree) are untested. **PRD milestone M6 — a real `workflow_dispatch` run, with the artifacts installed on real machines — is open and is the gate for that half.** Please do not read a green CI here as evidence that a bundle builds; nothing in CI runs `tauri build`.

## Gates

- `cargo fmt --check` — clean.
- `cargo clippy --workspace --all-targets --features e2e -- -D warnings` — **exit 0**, all three flags, desktop crate included.
- `cargo test-fast` — **3471/3471 passed**.
- `cargo test-e2e` — 9344 tests, 9339 passed, 5 failed, all five real-agent (LLM) tests. Re-run with identical code: **all 5 passed** (2 flagged SLOW, one 133s), so they are contention flakes from running alongside 9339 others — the tier rule 4 calls flaky-tolerant. Attribution is not a judgement call here: this branch changes zero files under `src/` and zero under `tests/`, so the e2e binaries are content-identical to `main`'s.
- Run with `--exclude dot-agent-deck-desktop`: that crate has no `e2e`-gated targets and its unit tests already passed in the fast tier. Excluding it avoids `LD_LIBRARY_PATH`, which this box needs for the desktop crate but which breaks nix's `node` — and the e2e tier spawns node-based agent CLIs.

## Also worth knowing

Both mandated gates are currently **unrunnable on a Linux devbox shell**: nix's glibc has no `ld.so.cache`, so `ldconfig -p` returns zero entries and system GTK/WebKit libraries are invisible to the loader — `cargo test-fast` dies with `libgdk-3.so.0: cannot open shared object file`. This has been true since `daf94f0` put `dot-agent-deck-desktop` in the workspace. Telling contributors to `apt install` does not fix it; the packages must be added to `devbox.json` as **nix** packages. Out of scope here, flagged for a separate issue.

Closes #740.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -491,6 +506,226 @@ jobs:
...
+          # Deliberately NOT `dot-agent-deck-*`: that is the pattern `finalize`
+          # selects on, and these must stay out of the CLI release's `dist/`.
+          name: desktop-bundle-${{ matrix.asset_suffix }}
+          path: dist-desktop/
+
+  # Attaches the desktop assets to the release that `finalize` already created.
+  # Separated from `desktop-bundle` so the upload happens once, after the
+  # release object exists, rather than racing per matrix leg.
+  desktop-publish:
+    needs: [prepare, finalize, desktop-bundle]
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply .github/workflows/release.yml:679]: Valid, and fixed in c23d3fc — thank you, this was a real defect sitting under twelve green checks.

You are right that `fail-fast: false` was doing less than it looked like. It prevents a failing leg from *cancelling* its sibling, so both legs run and both upload — but `needs:` on a matrix job resolves to the aggregate result, so one failed leg still skipped `desktop-publish` and discarded the surviving bundle *after* it had been built and uploaded. That is precisely the outcome the flag was added to prevent.

`desktop-publish` now gates on the release object existing rather than on every leg succeeding:

```yaml
if: ${{ !cancelled() && needs.finalize.result == 'success' && !inputs.skip_desktop }}
```

plus a guard step that fails loudly when no leg produced anything, and emits a per-platform `::warning::` when it publishes a partial set — a release carrying one platform where it should carry two should not be discovered from a user's bug report.

Your note that `release_workflow_wiring.rs` did not catch this was the sharper half of the finding: the `needs:` edge those tests asserted was correct, and it was the edge's *aggregate semantics* that were wrong. Two assertions added for both halves, and I mutation-tested them — restoring the old gate turns exactly one red.

One knock-on worth flagging: the fix tripped my own `desktop_jobs_do_not_swallow_their_own_failures`, which forbade `continue-on-error` anywhere in these jobs while the fix needs it on one *step* (so a partial matrix reaches the guard instead of dying on `download-artifact`'s less useful error). Right instinct, too blunt — narrowed to **job-level** `continue-on-error`, which is the one that actually suppresses a result. Narrowed, not deleted.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -491,6 +506,226 @@ jobs:
...
+            # it is one word here once it is pinned. See PRD #740 Decision 7.
+            bundles: deb
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v7
+      - uses: dtolnay/rust-toolchain@v1
+        with:
+          toolchain: 1.97.1
+          targets: ${{ matrix.target }}
+      - uses: Swatinem/rust-cache@v2
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply .github/workflows/release.yml:563]: Legitimate finding, and closely related to an argument this PR makes itself — Decision 7 drops AppImage precisely because its bundler executes shell scripts fetched from a mutable `master` ref inside the release path. So I am not disputing the principle. I am not fixing it in this PR, for two reasons.

**Degree.** `master` is a branch that moves continuously and can carry anything; `@v4` is a release tag a maintainer moves deliberately and only within a major. Related, not equivalent.

**Consistency, which is the deciding one.** `release.yml` is *uniformly* tag-pinned today — `actions/checkout@v7`, `actions/download-artifact@v8`, `actions/upload-artifact@v7`, `dtolnay/rust-toolchain@v1`, `softprops/action-gh-release@v3` — including every step of the `build` job that produces the CLI binaries users already install. Pinning only the new desktop job would leave one file with two conventions while the far more widely installed artifacts stayed unpinned, which is worse than either consistent state. Repo-wide, the 113 SHA pins that do exist are all in auto-generated agentic workflow lock files.

So the coherent change is repo-wide and belongs in its own PR, where the Renovate interaction can be thought through properly rather than as a side effect of a packaging change. Recorded in the PRD work log so it is not lost.

[the remark this answers is not in the document]
Comment (vfarcic) [reply .github/workflows/release.yml:563]: Tracked as #769, so resolving this thread here.

To be explicit about what was and was not done: the finding is accepted, and **nothing in this PR fixes it**. #769 covers pinning every `uses:` in `release.yml` to a commit SHA with a trailing `# vX.Y.Z` comment, which is the form Renovate keeps current.

The issue also records the two things that made deferring the right call rather than a dodge — that `release.yml` is uniformly tag-pinned today including the `build` job producing the CLI binaries users already install, and that `finalize` (which both creates the release and holds `HOMEBREW_TAP_TOKEN` and `SCOOP_BUCKET_TOKEN`) is a higher-value target than anything this PR added — and the counter-argument against it, which is that "the existing code is already unpinned" is exactly the reasoning by which nothing ever gets pinned. This PR adds two new entries to that gap, so it is a contributor to the problem, not a bystander.

**Review by prageethw — APPROVED** [review]
Reviewed the diff (release.yml's new desktop-bundle/desktop-publish jobs, xtask/linkage-check assertions, gitignore, prepare-sidecar.sh) and the PR body's claims against it. The job-topology guarantees (finalize never depends on desktop-bundle; desktop-publish gates on `!cancelled() && finalize.result == 'success'` rather than the matrix aggregate) are backed by mutation-tested static assertions in release_workflow_wiring.rs, not just prose. Both prior Greptile P1/P2 findings (partial-matrix-discard, AppImage mutable-ref supply chain) were fixed and confirmed addressed. All 12 CI checks green. Real `tauri build` output verified on Linux; macOS explicitly and honestly flagged as unverified pending PRD M6 — appropriately not claimed as covered by this green CI.
